### DISCUSSION TITLE
Alpine State Not Preserved After Livewire Rendering

### DISCUSSION THREAD

**Discussion by fredikaputra** [livewire/livewire#6172] [7. Bugs (archived)] [answered]
### Livewire version

v3.0.0-beta.6

### Which PHP version are you using?

PHP 8.2

### Repository URL

https://github.com/fredikaputra/livewire-rendering-bug

### Steps To Reproduce

1. Clone the project repository from `https://github.com/fredikaputra/livewire-rendering-bug`.
2. Install the required dependencies and start the project.
3. Open your web browser and navigate to http://localhost or the appropriate URL for accessing the project locally.
4. Locate the password input field on the page.
5. Type a password into the password input field.
6. Click the "Show Password" button (if applicable) to reveal the entered password.
7. While the password is visible, attempt to modify the password in the input field.
8. Observe that the input field type changes from "text" to "password" after Livewire performs rendering, causing the entered text to be hidden again.
9. The issue can be verified as the input field is not preserving the changes made to the password after Livewire performs rendering.

### Additional information

In the Livewire blade, the password input is defined as follows:

```html
<input type="password" placeholder="Your password" wire:model.live="password" x-ref="password_field">
```

The intended behavior is that when the "Show Password" button is clicked, the input type should change to "text" to reveal the password in plain text. However, an issue arises when attempting to change the password value. The live validation process triggers a re-rendering of the input element.

As a consequence of the re-rendering, the input type is reset to "password," causing the entered text to be hidden again instead of remaining visible as "text."

In summary, the problem is that after enabling the "Show Password" feature and modifying the password, the input type reverts back to "password" during the rendering process, leading to the loss of the temporarily revealed password.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by donkooijman** [comment, marked as answer]
Hi @fredikaputra! Thanks for sharing your code. I've taken a look at it, the server will always return 'password' for the type of the field and the Javascript you have to change this is not executed after Livewire made a round trip.

I've made some small adjustments that make it work, it might help you out. I used x-bind:type to dynamically set the type of the field. Take a look at this example to see how it works:

```
<div x-data="{ showPassword: false }">
    Your password is {{ $password }}
    <div>
        <input
            x-bind:type="showPassword ? `text` : `password`"
            placeholder="Your password"
            wire:model.live="password"
        >
        <div>
            @error('password')
                {{ $message }}
            @enderror
        </div>
    </div>
    <div>
        <button
            type="button"
            title="Show password"
            x-on:click="showPassword = ! showPassword"
            x-text="showPassword ? `Hide Password` : `Show Password`"
        ></button>
    </div>
</div>
```

**Comment by wizzymore** [comment]
> Alpine State Not Preserved After Livewire Rendering

This is really not the case here, Alpine is remembering the state of you component, but you don't have any. You are directly modifying the dom structure without using alpine hooks and a boolean in the alpine object that monitors if the input shows the password or not.

Everytime livewire will does a request to update the dom it will just go back to the initial form written in the `.blade.php` as there is nothing that tells that `input` element what type it is on render except the `type="password"` from the blade file.

**Reply by fredikaputra** [reply]
Oh, I see! You were right, and your explanation really helped me understand the issue. Thank you so much for taking the time to clarify things for me. I appreciate it!
